src/producer_processor_saver/Logic.java

```diff
@@ -0,0 +1,173 @@
+package producer_processor_saver;
+
+import java.util.LinkedList;
+import java.util.Queue;
+
+public class Logic {
+
+    /**
+     * Queue to emulate loading from disk.
+     */
+    private Queue<Integer> loadingQueue = new LinkedList<>();
+
+    /**
+     * Queue to emulate processed data ready for storing.
+     */
+    private Queue<String> processedQueue = new LinkedList<>();
+
+    private int maxCount;
+    private int maxLoadStorage;
+    private int maxProcessedStorage;
+    private int loadTime;
+    private int processingTime;
+    private int storeTime;
+
+    public Logic(int maxCount, int maxLoadStorage, int maxProcessedStorage, int loadTime, int processingTime,
+                 int storeTime) {
+        this.maxCount = maxCount;
+        this.maxLoadStorage = maxLoadStorage;
+        this.maxProcessedStorage = maxProcessedStorage;
+        this.loadTime = loadTime;
+        this.processingTime = processingTime;
+        this.storeTime = storeTime;
+    }
+
+    public void loadAll() {
+        int counter = 0;
+        boolean doLoad = true;
+        Integer loadedVal = null;
+        while (counter < maxCount) {
+            if (doLoad) {
+                loadedVal = load();
+            }
+
+            // insert in loadingQueue
+            synchronized(loadingQueue) {
+                if (loadingQueue.size() < maxLoadStorage) {
+                    loadingQueue.add(loadedVal);
+                    doLoad = true;
+                    counter++;
+                    loadingQueue.notify(); // notify processing
+                } else {
+                    try {
+                        loadingQueue.notify(); // notify processing
+                        loadingQueue.wait(); // release the lock
+                    } catch (InterruptedException e) {
+                        e.printStackTrace();
+                    } finally {
+                        doLoad = false;
+                    }
+                }
+            }
+        }
+    }
+
+    private int loadingCounter = 0;
+
+    private int load() {
+        System.out.println("Loading: " + loadingCounter);
+
+        try {
+            Thread.sleep(loadTime);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+
+        return loadingCounter++;
+    }
+
+    public void processAll() {
+        int counter = 0;
+        boolean doSave = false;
+        String processedResult = null;
+
+        while (counter < maxCount) {
+            // check if loaded part & processing part
+            Integer loadedData;
+            if (!doSave) {
```

Comment:
What is the purpose of `doSave`? Would anything break if you removed it and the `if`s that use it?